### ISSUE TITLE
NLU agent is slow to load

### ISSUE THREAD

**Issue by mouadino** [RasaHQ/rasa_nlu#1015]
**Rasa NLU version**: 0.11.0

**Operating system** (windows, osx, ...): OSX/Linux

**Content of model configuration file**:
```json
{
  "pipeline": "spacy_sklearn",
  "path" : "./resources/models/nlu",
  "data" : "./resources/data/nlu.md"
}
```

**Issue**:

This may be more of a question than a  bug, but we have been trying to deploy our small bot from [here](https://github.com/ThoughtWorksInc/twde-chatbot) to Heroku and it has been an exhausting exercise, first, we have to use the biggest type of Dyno since 2.5G wasn't enough, so now we habe 14G of RAM and our metrics show sometimes the bot use 6G of a small period of time) to run the training (sadly currently training is done in release time which we will try to fix later), but worst is that the time it takes to answer queries from our chatbot sometimes is in the 30 seconds range, after a bit of investigation and with the help with some profiling [1] we found out that the step that takes most of the time is the part we create an agent and load trained model from disk (it's around 15seconds, check below for the profile data), and while we worked around it to only happen at process start [2], it still wasn't enough b/c gunicorn (and even heroku worker queues) keep using processes for handling requests and reaping them after, IMHO relying on process life cycle is brittle and we would like to avoid, so I am wondering how do others deal with the cold start? sadly since our bot is in Python we have still the option to try using threads/greenthreads (instead of processes) for handling queries but I am still uneasy regarding this.

[1]  
Profiling data

```
13:29:06 web.1   |  PATH: '/endpoint'
13:29:06 web.1   |           7367407 function calls (7331638 primitive calls) in 19.162 seconds
13:29:06 web.1   |     Ordered by: internal time, call count
13:29:06 web.1   |     List reduced from 6749 to 30 due to restriction <30>
13:29:06 web.1   |     ncalls  tottime  percall  cumtime  percall filename:lineno(function)
13:29:06 web.1   |          1    3.651    3.651    3.651    3.651 vocab.pyx:450(lexemes_from_bytes)
13:29:06 web.1   |          1    3.509    3.509   11.746   11.746 vocab.pyx:367(from_disk)
13:29:06 web.1   |          1    2.886    2.886    3.624    3.624 /Users/mouad/.virtualenvs/twde-chatbot-kvnVSXOR/lib/python3.6/site-packages/spacy/_ml.py:244(link_vectors_to_models)
13:29:06 web.1   |    1349784    1.996    0.000    1.996    0.000 {built-in method numpy.core.multiarray.fromstring}
13:29:06 web.1   |    1352570    1.401    0.000    3.396    0.000 /Users/mouad/.virtualenvs/twde-chatbot-kvnVSXOR/lib/python3.6/site-packages/msgpack_numpy.py:68(decode)
13:29:06 web.1   |          1    1.132    1.132    4.518    4.518 {msgpack._unpacker.unpack}
13:29:06 web.1   |     145/73    0.728    0.005    0.908    0.012 {built-in method _imp.create_dynamic}
13:29:06 web.1   |    1343766    0.635    0.000    0.738    0.000 vocab.pyx:195(__iter__)
13:29:06 web.1   |       1128    0.357    0.000    0.357    0.000 {method 'read' of '_io.FileIO' objects}
13:29:06 web.1   |       1125    0.180    0.000    0.180    0.000 {built-in method marshal.loads}
13:29:06 web.1   |      17976    0.170    0.000    0.170    0.000 {built-in method posix.stat}
13:29:06 web.1   |          5    0.118    0.024    0.118    0.024 {built-in method _pywrap_tensorflow_internal.TF_Run}
13:29:06 web.1   |    1503/12    0.115    0.000    2.714    0.226 {built-in method builtins.exec}
13:29:06 web.1   |         45    0.112    0.002    0.112    0.002 {method 'uniform' of 'mtrand.RandomState' objects}
13:29:06 web.1   |        139    0.109    0.001    0.110    0.001 {built-in method io.open}
13:29:06 web.1   |    1348567    0.104    0.000    0.124    0.000 vocab.pyx:132(get_by_orth)
13:29:06 web.1   |        480    0.095    0.000    0.546    0.001 /usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/traceback.py:313(extract)
13:29:06 web.1   |       2470    0.070    0.000    0.092    0.000 vocab.pyx:228(make_fused_token)
13:29:06 web.1   |        434    0.063    0.000    0.067    0.000 /Users/mouad/.virtualenvs/twde-chatbot-kvnVSXOR/lib/python3.6/site-packages/tensorflow/python/framework/ops.py:2849(_extract_stack)
13:29:06 web.1   |       1131    0.052    0.000    0.409    0.000 <frozen importlib._bootstrap_external>:830(get_data)
13:29:06 web.1   |  2072/2053    0.043    0.000    0.082    0.000 {built-in method builtins.__build_class__}
13:29:06 web.1   |      50233    0.035    0.000    0.254    0.000 /usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/traceback.py:279(line)
13:29:06 web.1   |        179    0.033    0.000    0.033    0.000 {method 'readline' of '_io.BufferedReader' objects}
13:29:06 web.1   |          2    0.032    0.016    0.162    0.081 nn_parser.pyx:894(from_disk)
13:29:06 web.1   |        5/1    0.027    0.005   15.693   15.693 /Users/mouad/.virtualenvs/twde-chatbot-kvnVSXOR/lib/python3.6/site-packages/spacy/util.py:507(from_disk)
13:29:06 web.1   |      50231    0.026    0.000    0.242    0.000 /Users/mouad/.virtualenvs/twde-chatbot-kvnVSXOR/lib/python3.6/linecache.py:15(getline)
13:29:06 web.1   |      45190    0.024    0.000    0.025    0.000 {built-in method builtins.hasattr}
13:29:06 web.1   |       2234    0.022    0.000    0.126    0.000 <frozen importlib._bootstrap_external>:1233(find_spec)
13:29:06 web.1   |          1    0.022    0.022    0.090    0.090 pipeline.pyx:620(load_model)
13:29:06 web.1   |      50229    0.021    0.000    0.036    0.000 /Users/mouad/.virtualenvs/twde-chatbot-kvnVSXOR/lib/python3.6/linecache.py:147(lazycache)
```

[2] https://github.com/ThoughtWorksInc/twde-chatbot/blob/master/chatbot/messenger/google_chat_api.py#L15

**Comment by mouadino** [comment]
Related to this is an Agent instance (https://core.rasa.com/api/agent.html?highlight=agent) thread-safe?

**Comment by mouadino** [comment]
Small update, after spending most part of the day investigating more, we learned two things that I would like to share with the community:

1. We switched to use gevent with Gunicorn for the time being to avoid cold start of processes, since the default of Gunicorn is to use processes for handling workers, I am still unsure if an io-async flow is the best, since for my understanding, bots build on rasa may end up spending more time on CPU than IO.

2. It seems that Heroku may not be the best type of platforms to deploy such tools especially b/c of the slow start that Heroku impose on processes, where dynos that don't get used for some time Heroku put them to sleep, this lead the bot to start over for the unlucky ones that get to use the bot after a long silence, ofc we can set up external uptime system that ping the app often enough so it doesn't sleep, we will try that tomorrow and update here again.

FWIW concerning agent not being thread-safe, I would still love a hint about whether that's true or not.
